bindings/java/src/org/sleuthkit/datamodel/HostManager.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Sleuth Kit Data Model
+ *
+ * Copyright 2020 Basis Technology Corp.
+ * Contact: carrier <at> sleuthkit <dot> org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sleuthkit.datamodel;
+
+import com.google.common.base.Strings;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.sleuthkit.datamodel.SleuthkitCase.CaseDbConnection;
+import org.sleuthkit.datamodel.SleuthkitCase.CaseDbTransaction;
+
+/**
+ * Responsible for creating/updating/retrieving Hosts.
+ *
+ */
+public class HostManager {
+	
+	private static final Logger LOGGER = Logger.getLogger(HostManager.class.getName());
+
+	private final SleuthkitCase db;
+
+	/**
+	 * Construct a HostManager for the given SleuthkitCase.
+	 *
+	 * @param skCase The SleuthkitCase
+	 *
+	 */
+	HostManager(SleuthkitCase skCase) {
+		this.db = skCase;
+	}
+	
+	
+	/**
+	 * Get or create host with specified name.
+	 * 
+	 * @param name	Host name.
+	 * @param transaction Database transaction to use.
+	 * 
+	 * @return
+	 * @throws TskCoreException 
+	 */
+	Host getOrCreateHost(String name, CaseDbTransaction transaction) throws TskCoreException  {
+		
+		// must have a name
+		if (Strings.isNullOrEmpty(name) ) {
+			throw new IllegalArgumentException("Host name is required.");
+		}
+
+		CaseDbConnection connection = transaction.getConnection();
+
+		// First search for host by name
+		Optional<Host> host = getHost(name, connection);
+		if (host.isPresent()) {
+			return host.get();
+		}
+
+		// could'nt find it, create a new host
+		return createHost(name, connection);
+	}
+	
+	/**
+	 * Get host with given name. 
+	 * 
+	 * @param name Host name to look for. 
+	 * @param connection Database connection to use.
+	 * 
+	 * @return Optional with host.  Optional.empty if no matching host is found.
+	 * 
+	 * @throws TskCoreException 
+	 */
+	Optional<Host> getHost(String name, CaseDbConnection connection ) throws TskCoreException {
+		
+		String queryString = "SELECT * FROM tsk_hosts"
+							+ " WHERE LOWER(name) = LOWER('" + name + "')";
```

Comment:
The use of lower here seems inconsistent with the UNIQUE constraint on name, which will be allow case-sensitive variations.

Reply:
Host names are not case sensitive, I don't think we should force them to be.

In terms of implementation, this seemed like the best option - check ignoring the case, but then store what we got. (preserve case).

Reply:
That sounds logical, but then we had better lower case the host name on INSERT.

Reply:
This host name can also be whatever the user picks.  So, I think we should preserve whatever they type. IF they want a camel case name, we shoudl allow them too.  The future host addresses work is where actual host names will get stored.

Reply:
That's what was referring to when i said "store what we get". Windows is a case preserving OS.   Forcing host names to go lower case goes against that.